Src/Extern/SUNDIALS/AMReX_SUNMemory.cpp

```diff
@@ -0,0 +1,140 @@
+#include <AMReX.H>
+#include <AMReX_Gpu.H>
+#include <AMReX_SUNMemory.H>
+#if defined(AMREX_USE_HIP)
+#include <sunmemory/sunmemory_hip.h>
+#elif defined(AMREX_USE_CUDA)
+#include <sunmemory/sunmemory_cuda.h>
+#elif defined(AMREX_USE_DPCPP)
+#include <sunmemory/sunmemory_sycl.h>
+#endif
+
+namespace amrex {
+namespace sundials {
+
+namespace {
+  int Alloc(SUNMemoryHelper, SUNMemory* memptr, size_t memsize, SUNMemoryType mem_type)
+  {
+    SUNMemory mem = SUNMemoryNewEmpty();
+
+    mem->ptr = NULL;
+    mem->own = SUNTRUE;
+    mem->type = mem_type;
+
+    if (mem_type == SUNMEMTYPE_HOST) {
+      mem->ptr = The_Cpu_Arena()->alloc(memsize);
+    } else if (mem_type == SUNMEMTYPE_UVM) {
+      mem->ptr = The_Managed_Arena()->alloc(memsize);
+    } else if (mem_type == SUNMEMTYPE_DEVICE) {
+      mem->ptr = The_Device_Arena()->alloc(memsize);
```

Comment:
I suggest we use `The_Arena` when it's possible so that the memory is less fragmented.

```suggestion
    } else if (mem_type == SUNMEMTYPE_UVM) {
        if (The_Arena()->isManaged()) {
          mem->ptr = The_Arena_Arena()->alloc(memsize);
        } else if (The_Managed_Arena()->isManaged()) {
          mem->ptr = The_Managed_Arena()->alloc(memsize);
        } else {
          free(mem);
          return -1;
        }
    } else if (mem_type == SUNMEMTYPE_DEVICE) {
      mem->ptr = The_Arena()->alloc(memsize);
```